### DISCUSSION TITLE
FRC: Piece Gateway

### DISCUSSION THREAD

**Discussion by willscott** [filecoin-project/FIPs#711] [Enhancements - Technical]
We have had a protocol that has been offered by Boost for a while, and has been used by several of the FIL+ deal making programs. This is for a piece gateway - an HTTP endpoint exposed on SPs allowing retrieval of a full piece of data.

The full semantics of this gateway protocol have not been fully specified. It seems useful to codify them at a specification level, rather than point to a single implementation.

To that end, I'm proposing an FRC defining what we mean by a piece gateway.
I put together a first pass as https://github.com/filecoin-project/FIPs/pull/709 . It parallels in many ways the block gateway used by IPFS in structure, and expected HTTP semantics.

**Comment by willscott** [comment]
From subsequent conversations, I'm planning to extend the semantics here to also specify transfer of sealed pieces.

Currently,
`/piece/{piece cid}` is used for unsealed transfers. My inclination is that to keep similar semantics extending the piece gateway to support sealed data transfer, we would specify that one of:

1. `/sealed/{piece cid}`
2. `/piece/{piece cid}/sealed`
3. `/piece/{commR}`

Would be interested in what others would prefer here

**Reply by ribasushi** [reply]
As discussed in slack - only 3 is viable. A single CommP can (and does) have multiple replicas within the same SP. When downloading a sealed replica you must pull a specific CommR so you can find the correct seed to regenerate its SDR counterpart.

TLDR: all sealed replicas are unique, thus need to be addressed by a unique `CommR`

**Reply by willscott** [reply]
I've updated the language of the proposed FRC to include reference to supporting the fetch of `CommR`

**Comment by xinaxu** [comment]
Should we include unpadded piece retrieval in this FRC. This would help reducing number of bytes needed to be downloaded to verify the integrity as the downloader can can do the padding locally when checking commp.
A similar function like https://boost.filecoin.io/http-retrieval#retrieving-a-car-file but under /piece route

**Reply by willscott** [reply]
I believe the semantics here are for unpadded retrieval. That's a good call-out though, and we should make it clear whether the expected semantics are with or without padding

**Reply by ribasushi** [reply]
Note: "unpadded" piece payload is a heuristic. It only works if one assumes that the payload is a CAR, and the "when I see nulls I stop" guess doesn't misfire. While having this heuristic on by default is ok, there must be a flag to say "no, give me all the bytes you have on your end". This way folks who do store non-car-files in sectors ( which does include us with e.g. aggregate pieces ) will not get their world broken.

**Reply by willscott** [reply]
I assumed this was the fr32 padding. I guess we need to consider "zero-padding to piece size" and fr32 padding separately?

I think we don't need fr32. agree with @ribasushi that zero-padding needs a bit more consideration.
One thing i will say is that given any sort of transport-layer compression, e.g. gzip over http, the trailing zeros should not end up requiring any significant overhead on the wire, so i'm not sure we have a huge penalty in this transfer

**Reply by xinaxu** [reply]
@ribasushi Oh, I thought boost can record the size of imported file and use that to return non zero padded file, regardless it's a CAR or anything else.
@willscott the only issue with gzip is, it cannot be multithreaded.

**Reply by willscott** [reply]
multiple simultaneous requests for different ranges of the piece could probably be used if multi-threading efficiency was needed, right?

**Reply by ribasushi** [reply]
> I thought boost can record the size of imported file and use that to return non zero padded file

It can, but:
1) boost is not the only implementation of this
2) the hash function is defined over a power-of-2-sizes stream, there should be a mode to "give me the raw stream, no extra handlng"

> I assumed this was the fr32 padding. I guess we need to consider "zero-padding to piece size" and fr32 padding separately?

I avoid this clash by referring to:
- "get the payload to the `2^N` 127-byte leaves" as `padding`
- and "get every 127 bytes of payload to 4 BLS-compatible 254-bit leaves" as `fr32 expansion`

**Reply by xinaxu** [reply]
@ribasushi with majority using boost, I suggest we have that optimization ready for trailing zeroes.
@willscott It's not possible. I did some analysis when I was trying to use booster-http to distribute car files to other miners long time ago. That's because gzip content encoding means the compression needs to happen for the whole file and does not work with ranges request.
if a client sends a range request for a resource that is being gzip-compressed on the fly, Nginx typically ignores the range request and sends the entire compressed resource.
To allow gzip compression for any ranges, one needs to use transfer-encoding instead of content encoding. This is what you described I believe, however both server and client support is lacking.
Another way to serve gzip compression with ranges header is to have the file zipped first in which case nginx will look for the zipped file first.
However when using nginx with boost, it is not serving static files but proxy the response instead.

**Reply by ribasushi** [reply]
> I suggest we have that optimization ready for trailing zeroes.

To rephrase: pieces prepared via [go-data-segment](https://pkg.go.dev/github.com/filecoin-project/go-data-segment/datasegment#Aggregate) do not have trailing zeroes in the 1st place.

**Reply by Kubuxu** [reply]
> To rephrase: pieces prepared via [go-data-segment](https://pkg.go.dev/github.com/filecoin-project/go-data-segment/datasegment#Aggregate) do not have trailing zeroes in the 1st place.

I don't think this is completely correct. The `PieceInfo.Size.Validate()` call for sub-pieces was missing, I added it some time ago. It is not 100% necessary, go-data-segment can work with non-power-of-two pieces, but the data type it currently uses suggests that these sizes should be a power of two.
As I said, it is only an external API assumption. Internally it can work with any piece size (same as the verifier side as long as the client passes the non-zero padded size).

**Reply by Kubuxu** [reply]
I misunderstood the discussion here. This is about fetching the whole deal/piece, not sub-pieces.

In that case, I would suggest a much simpler solution: requiring support for compression for these HTTP endpoints. Large runs of zeros will compress very well, even with the simplest deflate HTTP compression.

**Reply by willscott** [reply]
I've noted that transport compression [may be used](https://github.com/filecoin-project/FIPs/pull/709/files#diff-0951ff57060b1bf09b7f0ef35aa697778829927b04263bc4b391fe127a560fa2R172) - are you asking for a definition in the specification that server implementations must support specific forms of http transport compression?

**Reply by ribasushi** [reply]
@willscott given how common significant runs-of-`\x00` will be in this context, we probably could throw a `SHOULD` in the spec... Not feeling strongly either way though

**Reply by xinaxu** [reply]
Reiterating my point, correct me if I'm wrong, there is no client supporting gzip transfer encoding, i.e. aria2c only supports gzip content compression, but not transfer compression.
How about a custom response header that indicates known range of non zeroes? So users can HEAD first and then do range request.

**Reply by Kubuxu** [reply]
@willscott It should be Content-Encoding (so content compression). Transfer compression is widely unsupported, as @xinaxu mentions.

We could go as far as requiring that compliant servers MUST support at least `gzip` content compression (and `identity` for no compression), and clients SHOULD support and use `gzip` in their `Accept-Encoding` header.

**Reply by xinaxu** [reply]
Thanks, using gzip content compression is good enough to some extent, though I'd complain that it won't work with multithreading - because it can't compress from the middle. 
Not a blocker but in pratice when I distribute CAR files to my partner SPs, all of them are using multithreaded downloading clients which is almost a requirement for transfer cross continents.
